### PR TITLE
add garbase collect by rowsetid

### PR DESCRIPTION
**Pull request by kangpinghuang**

### CODE REVIEW

File: be/src/olap/storage_engine.cpp [hunk 1]
```diff
@@ -944,6 +944,20 @@ OLAPStatus StorageEngine::execute_task(EngineTask* task) {
     }
 }
 
+bool StorageEngine::check_path_in_unused_rowsets(const string& path) {
```
Comment (yiguolei) [inline be/src/olap/storage_engine.cpp]: Add some comment to describe the path strategy

File: be/src/olap/data_dir.cpp [hunk 2]
```diff
@@ -932,6 +932,52 @@ void DataDir::perform_path_gc() {
...
+        if (!is_valid) {
+            LOG(WARNING) << "unknown path:" << path;
+            continue;
+        }
+        if (tablet_id > 0 && schema_hash > 0) {
+            // tablet schema hash path or rowset file path
+            // gc thread should get tablet include deleted tablet
+            // or it will delete rowset file before tablet is garbage collected
+            RowsetId rowset_id = -1;
+            bool is_rowset_file = _tablet_manager->get_rowset_id_from_path(path, &rowset_id);
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp]: get_rowset_id_from_path should in rowset class?

File: be/src/olap/tablet.h [hunk 3]
```diff
@@ -248,6 +252,7 @@ class Tablet : public std::enable_shared_from_this<Tablet> {
 
     std::atomic<bool> _is_bad;   // if this tablet is broken, set to true. default is false
     std::atomic<int64_t> _last_compaction_failure_time; // timestamp of last compaction failure
+    RowsetId _start_rowset_id;
```
Comment (yiguolei) [inline be/src/olap/tablet.h]: start_rowset_id is useless?

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -839,6 +839,12 @@ bool Tablet::check_path(const std::string& path_to_check) {
             return true;
         }
     }
+    for (auto& inc_version_rowset : _inc_rs_version_map) {
+        bool ret = inc_version_rowset.second->check_path(path_to_check);
+        if (ret) {
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp]: maybe only check rowset id not the full path?

File: be/src/olap/data_dir.cpp [hunk 5]
```diff
@@ -932,6 +932,52 @@ void DataDir::perform_path_gc() {
...
+            if (is_rowset_file) {
+                TabletSharedPtr tablet = _tablet_manager->get_tablet(tablet_id, schema_hash, true);
+                if (tablet != nullptr) {
+                    bool valid = tablet->check_path(path);
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp]: should sleep in path scan method to prevent io 100%

File: be/src/olap/data_dir.cpp [hunk 6]
```diff
@@ -932,6 +932,52 @@ void DataDir::perform_path_gc() {
...
+                    if (!valid) {
+                        // if the rowset id is less than tablet's initial end rowset id
+                        // and the path is not in unused_rowsets, delete the path.
+                        if (rowset_id < tablet->initial_end_rowset_id()
+                                && !StorageEngine::instance()->check_path_in_unused_rowsets(path)) {
+                            _process_garbage_path(path);
+                        }
+                    }
+                }
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp]: sleep every 1000 path?

Comment (kangpinghuang) [reply be/src/olap/data_dir.cpp]: every path_gc_check_step paths, I will sleep 10ms, it is a config.

File: be/src/olap/tablet.cpp (hunk 4 above)
Comment (kangpinghuang) [reply be/src/olap/tablet.cpp]: OK， I will add check_rowset_id to do this

File: be/src/olap/tablet.h (hunk 3 above)
Comment (kangpinghuang) [reply be/src/olap/tablet.h]: OK, I will remove it.

File: be/src/olap/data_dir.cpp (hunk 2 above)
Comment (kangpinghuang) [reply be/src/olap/data_dir.cpp]: I don't think so, because this is a path pattern. And more here you can not get a rowset to get the rowsetid

**Review by yiguolei — APPROVED** [review]
LGTM

File: be/src/olap/tablet.cpp [hunk 7]
```diff
@@ -839,6 +839,30 @@ bool Tablet::check_path(const std::string& path_to_check) {
...
+            return true;
+        }
+    }
+    return false;
+}
+
+bool Tablet::check_rowset_id(RowsetId rowset_id) {
+    bool ret = false;
```
Comment (chaoyli) [inline be/src/olap/tablet.cpp]: _rs_version_map will be modified when you call this function.
A readlock is necessary.

File: be/src/olap/tablet.cpp [hunk 8]
```diff
@@ -839,6 +839,30 @@ bool Tablet::check_path(const std::string& path_to_check) {
...
+    for (auto& version_rowset : _rs_version_map) {
+        ret = version_rowset.second->rowset_id() == rowset_id;
```
Comment (chaoyli) [inline be/src/olap/tablet.cpp]: if (version_rowset.second->rowset_id() == rowset_id) {
    return true;
}

File: be/src/olap/tablet.cpp [hunk 9]
```diff
@@ -839,6 +839,30 @@ bool Tablet::check_path(const std::string& path_to_check) {
...
+        if (ret) {
+            return true;
+        }
+    }
+
+    for (auto& inc_version_rowset : _inc_rs_version_map) {
+        ret = inc_version_rowset.second->rowset_id() == rowset_id;
```
Comment (chaoyli) [inline be/src/olap/tablet.cpp]: if (inc_version_rowset.second->rowset_id() == rowset_id) {
   return true;
}

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**